pkg/controller/jobs/jobset/jobset_controller.go

```diff
@@ -0,0 +1,223 @@
+	j.Spec.Suspend = pointer.Bool(true)
+}
+
+func (j *JobSet) GetGVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (j *JobSet) PodSets() []kueue.PodSet {
+	podSets := make([]kueue.PodSet, len(j.Spec.ReplicatedJobs))
+	for index, replicatedJob := range j.Spec.ReplicatedJobs {
+		podSets[index] = kueue.PodSet{
+			Name:     replicatedJob.Name,
+			Template: *replicatedJob.Template.Spec.Template.DeepCopy(),
+			Count:    podsCount(&replicatedJob),
+		}
+	}
+	return podSets
+}
+
+func (j *JobSet) RunWithPodSetsInfo(podSetInfos []jobframework.PodSetInfo) {
+	j.Spec.Suspend = pointer.Bool(false)
+	if len(podSetInfos) != len(j.Spec.ReplicatedJobs) {
+		return
+	}
+
+	// If there are Jobs already created by the JobSet, their node selectors will be updated by the JobSet controller
+	// before unsuspending the individual Jobs.
+	for index := range j.Spec.ReplicatedJobs {
+		templateSpec := &j.Spec.ReplicatedJobs[index].Template.Spec.Template.Spec
+		templateSpec.NodeSelector = maps.MergeKeepFirst(podSetInfos[index].NodeSelector, templateSpec.NodeSelector)
+	}
+}
+
+func (j *JobSet) RestorePodSetsInfo(podSetInfos []jobframework.PodSetInfo) {
+	if len(podSetInfos) == 0 {
+		return
+	}
+	for index := range j.Spec.ReplicatedJobs {
+		if equality.Semantic.DeepEqual(j.Spec.ReplicatedJobs[index].Template.Spec.Template.Spec.NodeSelector, podSetInfos[index].NodeSelector) {
+			continue
+		}
+		j.Spec.ReplicatedJobs[index].Template.Spec.Template.Spec.NodeSelector = maps.Clone(podSetInfos[index].NodeSelector)
+	}
+}
+
+func (j *JobSet) Finished() (metav1.Condition, bool) {
+	if apimeta.IsStatusConditionTrue(j.Status.Conditions, string(jobsetapi.JobSetCompleted)) {
+		condition := metav1.Condition{
+			Type:    kueue.WorkloadFinished,
+			Status:  metav1.ConditionTrue,
+			Reason:  "JobSetFinished",
+			Message: "JobSet finished successfully",
+		}
+		return condition, true
+	}
+	if apimeta.IsStatusConditionTrue(j.Status.Conditions, string(jobsetapi.JobSetFailed)) {
+		condition := metav1.Condition{
+			Type:    kueue.WorkloadFinished,
+			Status:  metav1.ConditionTrue,
+			Reason:  "JobSetFinished",
+			Message: "JobSet failed",
+		}
+		return condition, true
+	}
+	return metav1.Condition{}, false
+}
+
+func (j *JobSet) EquivalentToWorkload(wl kueue.Workload) bool {
+	podSets := wl.Spec.PodSets
+	if len(podSets) != len(j.Spec.ReplicatedJobs) {
+		return false
+	}
+
+	for index := range j.Spec.ReplicatedJobs {
+		if wl.Spec.PodSets[index].Count != podsCount(&j.Spec.ReplicatedJobs[index]) {
+			return false
+		}
+
+		jobPodSpec := &j.Spec.ReplicatedJobs[index].Template.Spec.Template.Spec
+		if !equality.Semantic.DeepEqual(jobPodSpec.InitContainers, podSets[index].Template.Spec.InitContainers) {
+			return false
+		}
+		if !equality.Semantic.DeepEqual(jobPodSpec.Containers, podSets[index].Template.Spec.Containers) {
+			return false
+		}
+	}
+	return true
+}
+
+func (j *JobSet) PriorityClass() string {
+	for _, replicatedJob := range j.Spec.ReplicatedJobs {
```

Comment:
Another thought for follow up, but we could be getting the priority from `PodSets()`

Reply:
it might not be tha case for all job types  ... maybe  make `PriorityClass` part of an optional interface and fallback on podset search.

Reply:
#921